### PR TITLE
Updating auto subscribe function

### PR DESCRIPTION
**Pull request by SumoSourabh**
adding license file, updating code and template to use destiantion arn along with rolearn

### CODE REVIEW

File: loggroup-lambda-connector/src/loggroup-lambda-connector.js [hunk 1]
```diff
@@ -5,13 +5,24 @@ async function sleep(waitTimeInMs) {
...
 }
 
-function subscribeToLambda(lambdaLogGroupName, lambdaArn, errorHandler) {
-    var params = {
-        destinationArn: lambdaArn,
-        filterName: 'SumoLGLBDFilter',
-        filterPattern: '',
-        logGroupName: lambdaLogGroupName
-    };
+function subscribeToLambda(lambdaLogGroupName, destinationArn, roleArn, errorHandler) {
```
Comment (himanshu219) [inline loggroup-lambda-connector/src/loggroup-lambda-connector.js:8]: the function name is subscribeToLambda probably we can change it to subscribeLogGroupToService

**Review by himanshu219 — CHANGES_REQUESTED** [review]
actually all the sam related files should in sam folder can we move it to sam. I guess I forgot to commit that file.  Rest of the changes looks good to me.

**Review by himanshu219 — APPROVED** [review]

**Review by himanshu219 — CHANGES_REQUESTED** [review]

File: loggroup-lambda-connector/src/loggroup-lambda-connector.js [hunk 2]
```diff
@@ -50,38 +53,41 @@ function filterLogGroups(event, logGroupRegex) {
...
-            createSubscriptionFilter(logGroupName, destinationArn, roleArn, (function(inner_logGroupName) { return function (err, data) {
-                if (err) {
-                    console.log("Error in subscribing", inner_logGroupName, err);
-                } else {
-                    console.log("Successfully subscribed logGroup: ", inner_logGroupName);
-                }
-            };})(logGroupName));
-        } else {
+    const failedLogGroupNames = [];
+    await logGroups.reduce(async (previousPromise, nextLogGroup) => {
```
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/src/loggroup-lambda-connector.js:61]: Nop, I tried and it was working async.

File: loggroup-lambda-connector/src/loggroup-lambda-connector.js [hunk 3]
```diff
@@ -16,15 +13,21 @@ function createSubscriptionFilter(lambdaLogGroupName, destinationArn, roleArn, e
...
             filterPattern: '',
             logGroupName: lambdaLogGroupName,
             roleArn: roleArn
         };
     }
 
     // handle case where subscription filter exists/case where loggroup generated by target lambda
-    cwl.putSubscriptionFilter(params, errorHandler);
+    try {
+        await util.promisify(cwl.putSubscriptionFilter.bind(cwl))(params);
```
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/src/loggroup-lambda-connector.js:25]: automatically available

File: loggroup-lambda-connector/sam/packaged.yaml [hunk 4]
```diff
@@ -90,18 +150,78 @@ Resources:
...
+            Ref: SumoLogGroupLambdaConnector
+      Policies:
+      - Statement:
+        - Sid: InvokePolicy
+          Effect: Allow
+          Action:
+          - lambda:InvokeFunction
+          Resource:
+          - Fn::Sub: arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:${SumoLogGroupLambdaConnector}
+  primerinvoke:
```
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/sam/packaged.yaml:210]: Sure

File: loggroup-lambda-connector/src/loggroup-lambda-connector.js [hunk 5]
```diff
@@ -39,37 +53,41 @@ function filterLogGroups(event, logGroupRegex) {
...
+                if (err && err.code == "ThrottlingException") {
+                    failedLogGroupNames.push({logGroupName: logGroupName});
+                }
+            });
         }
+    }, Promise.resolve());
+
+    if (counter < 4 && failedLogGroupNames.length > 0) {
+        console.log("Retrying Subscription for Failed Log Groups due to throttling with counter number as " + counter);
+        await subscribeExistingLogGroups(failedLogGroupNames, counter + 1);
```
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/src/loggroup-lambda-connector.js:78]: Sure

File: loggroup-lambda-connector/sam/template.yaml [hunk 6]
```diff
@@ -74,29 +116,107 @@ Resources:
...
+              Action:
+                - lambda:InvokeFunction
+              Resource:
+                - !Sub 'arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:${SumoLogGroupLambdaConnector}'
+
+  primerinvoke:
+    Type: AWS::CloudFormation::CustomResource
+    Version: "1.0"
+    Condition: invoke_existing
+    DeletionPolicy : "Retain"
```
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/sam/template.yaml:210]: The lambda is invoked again on deletion, which is not required.

File: loggroup-lambda-connector/sam/template.yaml [hunk 7]
```diff
@@ -74,29 +116,107 @@ Resources:
...
+      SourceAccount: !Ref AWS::AccountId
+      SourceArn: !Sub 'arn:aws:logs:${AWS::Region}:${AWS::AccountId}:log-group:*:*'
+
+  SumoLogGroupExistingLambdaConnector:
+    Type: AWS::Serverless::Function
+    Condition: invoke_existing
+    Properties:
+      InlineCode: |
+        var aws = require('aws-sdk');
+        var response = require('cfn-response');
```
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/sam/template.yaml:169]: Yes, but it is only available using inline code. If we use it via src, then it is not available.

**Comment by himanshu219** [comment]
LGTM

File: loggroup-lambda-connector/src/loggroup-lambda-connector.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/src/loggroup-lambda-connector.js]: Changes done

File: loggroup-lambda-connector/sam/packaged.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (SumoSourabh) [reply loggroup-lambda-connector/sam/packaged.yaml:210]: Changes done.

**Pull request merged**